### DISCUSSION TITLE
How to collect invalid inputs?

### DISCUSSION THREAD

**Discussion by avsaase** [pest-parser/pest#909] [Q&A]
I'm trying to write a parser for cron expressions that collects invalid inputs instead of failing to parse the expression altogether. For example, when parsing `1,5- 4 * * *`, I want to record the valid `1` and invalid `5-` entries for the minutes field, the valid `4` entry for the hours and the valid `*` for the remaining fields.

This is the grammar I put together:
```
all     = { "*" }
number  = { (ASCII_DIGIT)+ }
range   = { number ~ "-" ~ number }
step    = { (all | range) ~ "/" ~ number }
invalid = { ANY+ }
entry   = { (step | range | number | all | invalid) }
field   = { entry ~ ("," ~ entry)* }
ws      = { (" ")+ }
cron    = { SOI ~ #minutes = field ~ ws ~ #hours = field ~ ws ~ #days_of_month = field ~ ws ~ #months = field ~ ws ~ #days_of_week = field ~ EOI }
```
The problem is that the `5` gets eagerly parsed by `number` and the `-` is left for the next rule which then fails. Similarly, when parsing `-1 4 * * *`, the `invalid` rule captures the full input and the parser fails because there's nothing left to parse. Is there some way to look head in the `entry` rule to make sure the wrong rule doesn't capture part of the input? I feel like using the stack could be useful here but I have a hard time wrapping my head around how to actually use it.

**Comment by skittishdev** [comment]
Does this work?

```
WHITESPACE = _{ " " }
all     = { "*" }
number  = { (ASCII_DIGIT)+ }
range   = { number ~ "-" ~ number }
step    = { (all | range) ~ "/" ~ number }
entry   = { (step | range | number | all ) }
field   = { entry ~ (",")* }
cron    = { SOI ~ #minutes = field ~ #hours = field ~ #days_of_month = field ~ #months = field ~ #days_of_week = field ~ EOI }
```
